api/handlers_datasets.go

```diff
@@ -272,11 +272,54 @@ func (s *server) DatasetDeleteHandler(w http.ResponseWriter, r *http.Request) {
 	w = LogWriter{w}
 	vars := mux.Vars(r)
 	account := vars["account"]
-	dataset := vars["id"]
+	id := vars["id"]
+
+	service, ok := s.datasetServices[account]
+	if !ok {
+		msg := fmt.Sprintf("account not found: %s", account)
+		handleError(w, apierror.New(apierror.ErrNotFound, msg, nil))
+		return
+	}
+
+	log.Debugf("deleting data set %s for account %s", id, account)
+
+	// get metadata from repository
+	metadataOutput, err := service.MetadataRepository.Get(r.Context(), account, id)
+	if err != nil {
+		handleError(w, err)
+		return
+	}
+
+	dataRepo, ok := service.DataRepository[metadataOutput.DataStorage]
+	if !ok {
+		msg := fmt.Sprintf("requested data repository type not supported for this account: %s", metadataOutput.DataStorage)
+		handleError(w, apierror.New(apierror.ErrBadRequest, msg, nil))
+		return
+	}
+
+	// revoke access to the data repository (cleans up associated IAM policies)
+	if err := dataRepo.RevokeAccess(r.Context(), id); err != nil {
+		msg := fmt.Sprintf("failed to revoke data repository access for dataset %s", id)
+		handleError(w, apierror.New(apierror.ErrInternalError, msg, err))
+		return
+	}
+
+	// delete metadata
+	// TODO: Maybe we just mark the dataset as deleted in the metadata
+	if err = service.MetadataRepository.Delete(r.Context(), account, id); err != nil {
+		msg := fmt.Sprintf("failed to delete metadata for dataset %s", id)
+		handleError(w, apierror.New(apierror.ErrInternalError, msg, err))
+		return
+	}
 
-	log.Debugf("deleting data set %s for account %s", dataset, account)
+	// delete data repository (needs to be empty)
+	// Currently, this will only delete an empty bucket, or move on otherwise
+	if err = dataRepo.Delete(r.Context(), id); err != nil {
+		msg := fmt.Sprintf("failed to delete data repository for dataset %s", id)
+		log.Warn(msg)
+	}
```

Comment:
do you actually want to continue if this fails?  maybe move it above the metadata delete.

Reply:
i should probably move it first, then the RevokeAccess, and lastly the metadata delete
but this means the DELETE will fail if a bucket is not empty, which is fine for now.
eventually, we might want to have a soft delete (default) which only marks a dataset as deleted, and a hard delete (with a special query param) which actually deletes (like the current behavior), although even then i wouldn't want to purge the bucket automatically

Reply:
actually, this order makes more sense: 
  1) delete data repository
  2) delete metadata
  3) cleanup IAM (this may just log on error since the dataset is practically gone at this point)